### PR TITLE
Added muon SFs

### PR DESCRIPTION
**Pull request by gmadigan**

### CODE REVIEW

File: NTupleAnalyzer_nanoAOD.py [hunk 1]
```diff
@@ -159,6 +159,14 @@
 #_kinematicvariables += ['passWptCut','passZptCut','WorZSystemPt']
 #_kinematicvariables += ['WSystemPt','ZSystemPt']
 _kinematicvariables += ['matchedLQ']
+_kinematicvariables += ['recoSFMu1','recoSFupMu1','recoSFdownMu1']
```
Comment (DavidMorse) [inline NTupleAnalyzer_nanoAOD.py]: Please change naming convention to the following, e.g. this line would be

_kinematicvariables += ['mu1recoSF','mu1recoSFup','mu1recoSFdown']

etc. for mu2 and other SF

File: NTupleAnalyzer_nanoAOD.py [hunk 2]
```diff
@@ -881,6 +889,204 @@ def PropagatePTChangeToMET(met,original_object,varied_object):
 # 			taus.append(ThisTau)
 # 	return taus
 
+def getMuonSF(_pt,_eta,_year,):
```
Comment (DavidMorse) [inline NTupleAnalyzer_nanoAOD.py]: Why is there a comma after year?

File: NTupleAnalyzer_nanoAOD.py [hunk 3]
```diff
@@ -881,6 +889,204 @@ def PropagatePTChangeToMET(met,original_object,varied_object):
...
+	# Check if muon is in eta range, set list of SFs and errors (syst+stat), sorted by pt.
+	if _year=='2016':
+		# Muon reco SFs
+		# https://twiki.cern.ch/twiki/bin/view/CMS/HighPtMuonReferenceRun2
+		if abseta>=0.00 and abseta<=1.60: recoSFbyPt = [[0.9914, 0.0008],[0.9936, 0.0009],[0.993, 0.001],[0.993, 0.002],[0.990, 0.004],[0.990, 0.003],[0.989, 0.004],[0.8, 0.3]]
+		elif abseta>1.60 and abseta<=2.40: recoSFbyPt = [[1.00, 0.00],[0.993, 0.001],[0.991, 0.001],[0.985, 0.001],[0.981, 0.002],[0.979, 0.004],[0.978, 0.005],[0.9, 0.2]]
+		# Muon high pt identification scale factors
+		# Periods B,C,D,E,F: https://gitlab.cern.ch/cms-muonPOG/MuonReferenceEfficiencies/blob/master/EfficienciesStudies/2016_legacy_rereco/systematic/RunBCDEF_SF_ID.json 
+		# Periods G,H: https://gitlab.cern.ch/cms-muonPOG/MuonReferenceEfficiencies/blob/master/EfficienciesStudies/2016_legacy_rereco/systematic/RunGH_SF_ID.json
+		# Each run was properly weighted by luminosoty and combined into a single SF
```
Comment (DavidMorse) [inline NTupleAnalyzer_nanoAOD.py]: luminosity :)

File: NTupleAnalyzer_nanoAOD.py [hunk 4]
```diff
@@ -881,6 +889,204 @@ def PropagatePTChangeToMET(met,original_object,varied_object):
...
+		if _eta>1.70 and _eta<=2.00: highPtIdSFbyPt = [[0.985, 0.005],[0.986, 0.002],[0.988, 0.001],[0.989, 0.001],[0.986, 0.002],[0.984, 0.004]]
+		if _eta>2.00 and _eta<=2.10: highPtIdSFbyPt = [[0.983, 0.009],[0.983, 0.003],[0.987, 0.002],[0.989, 0.002],[0.986, 0.004],[0.995, 0.008]]
+		if _eta>2.10 and _eta<=2.20: highPtIdSFbyPt = [[0.983, 0.006],[0.975, 0.004],[0.966, 0.003],[0.971, 0.003],[0.965, 0.005],[0.96, 0.01]]
+		if _eta>2.20 and _eta<=2.30: highPtIdSFbyPt = [[1.0, 0.01],[0.993, 0.004],[0.994, 0.003],[0.989, 0.004],[0.981, 0.005],[0.99, 0.01]]
+		if _eta>2.30 and _eta<=2.40: highPtIdSFbyPt = [[1.0, 0.01],[0.975, 0.006],[0.982, 0.003],[0.991, 0.004],[0.986, 0.008],[1.0, 0.01]]		
+		# Muon relative track isolation scale factors
+		# Periods B,C,D,E,F: * https://gitlab.cern.ch/cms-muonPOG/MuonReferenceEfficiencies/blob/master/EfficienciesStudies/2016_legacy_rereco/systematic/RunBCDEF_SF_ISO.json
+		# Periods G,H: https://gitlab.cern.ch/cms-muonPOG/MuonReferenceEfficiencies/blob/master/EfficienciesStudies/2016_legacy_rereco/systematic/RunGH_SF_ISO.json
+		# Each run was properly weighted by luminosoty and combined into a single SF
```
Comment (DavidMorse) [inline NTupleAnalyzer_nanoAOD.py]: luminosity :)

File: NTupleAnalyzer_nanoAOD.py [hunk 5]
```diff
@@ -881,6 +889,204 @@ def PropagatePTChangeToMET(met,original_object,varied_object):
...
+		# High level trigger scale factors
+		# https://twiki.cern.ch/twiki/bin/view/CMS/HighPtMuonReferenceRun2
+		if abseta>=0.00 and abseta<=0.90: hltSFbyPt = [[0.980, 0.006],[0.982, 0.004],[0.976, 0.001],[0.980, 0.002],[0.978, 0.004],[0.973, 0.009],[0.957, 0.03]]
+		elif abseta>0.900 and abseta<=1.20: hltSFbyPt = [[0.982, 0.01],[0.966, 0.005],[0.967, 0.002],[0.964, 0.004],[0.960, 0.009],[0.985, 0.02],[0.988, 0.06]]
+		elif abseta>1.20 and abseta<2.10: hltSFbyPt = [[0.983, 0.01],[1.012, 0.008],[1.002, 0.003],[1.004, 0.004],[1.009, 0.009],[0.993, 0.01],[1.062, 0.04]]
+		elif abseta>2.10 and abseta<2.40: hltSFbyPt = [[0.987, 0.03],[0.952, 0.02],[0.999, 0.008],[1.003, 0.01],[1.024, 0.02],[0.993, 0.05],[1.070, 0.2]]
+
+	# Check muon pt and select SF from appropriate pt bin. Different SFs (e.g., reco vs hlt) have different pt ranges/binning. 
+	# Maps ensure double checking of pt bins do not occur (speeds up loop) for different SF types
+	lowBndMap1 = {0:0, 1:2, 2:3, 3:4, 4:5, 5:5, 6:6, 7:6}
```
Comment (DavidMorse) [inline NTupleAnalyzer_nanoAOD.py:1038]: Please add one more comment:

# This should be checked carefully if any binning changes

File: NTupleAnalyzer_nanoAOD.py [hunk 6]
```diff
@@ -1673,6 +1878,11 @@ def FullKinematicCalculation(T,variation):
 	_DPhiu2j1 = abs(muons[1].DeltaPhi(jets[0]))
 	_DPhiu2j2 = abs(muons[1].DeltaPhi(jets[1]))
 
+	year = '2016' #Fixme need to integrate year option
```
Comment (DavidMorse) [inline NTupleAnalyzer_nanoAOD.py:1882]: this is fine for now, but we need to make this a global option eventually

**Review by DavidMorse — CHANGES_REQUESTED** [review]

File: NTupleAnalyzer_nanoAOD.py (hunk 6 above)
Comment (gmadigan) [reply NTupleAnalyzer_nanoAOD.py:1882]: I suppose the samples will be sorted into directories by year at some stage--could we just pull the year from the path? Or, would it be better to specify as a command line option?

File: NTupleAnalyzer_nanoAOD.py (hunk 5 above)
Comment (gmadigan) [reply NTupleAnalyzer_nanoAOD.py:1038]: I take it you mean if we change the scale factors we use (might change pt binning), check the binning? In other words, leave it as is for now.

File: NTupleAnalyzer_nanoAOD.py (hunk 3 above)
Comment (gmadigan) [reply NTupleAnalyzer_nanoAOD.py]: Oops!

File: NTupleAnalyzer_nanoAOD.py (hunk 2 above)
Comment (gmadigan) [reply NTupleAnalyzer_nanoAOD.py]: Typo

File: NTupleAnalyzer_nanoAOD.py (hunk 5 above)
Comment (DavidMorse) [reply NTupleAnalyzer_nanoAOD.py:1038]: Yes don't change the code, just add a warning comment "# This should....."

File: NTupleAnalyzer_nanoAOD.py (hunk 6 above)
Comment (DavidMorse) [reply NTupleAnalyzer_nanoAOD.py:1882]: We can decide later, not a problem. Not sure yet which is best...

File: NTupleAnalyzer_nanoAOD.py (hunk 5 above)
Comment (gmadigan) [reply NTupleAnalyzer_nanoAOD.py:1038]: Ok, will do

**Comment by gmadigan** [comment]
Fixed changes proposed by Dave.

**Review by DavidMorse — APPROVED** [review]

**Pull request merged**